tools/lkl/lib/virtio.c

```diff
@@ -544,24 +546,79 @@ int virtio_dev_setup(struct virtio_dev *dev, int queues, int num_max)
 		return -LKL_ENOMEM;
 	}
 
-	avail = sizeof(lkl_virtio_devs) - (devs - lkl_virtio_devs);
-	num_bytes = snprintf(devs, avail, " virtio_mmio.device=%d@0x%lx:%d",
-			 mmio_size, (uintptr_t)dev->base, dev->irq);
-	if (num_bytes < 0 || num_bytes >= avail) {
-		lkl_put_irq(dev->irq, "virtio");
-		unregister_iomem(dev->base);
-		lkl_host_ops.mem_free(dev->queue);
-		return -LKL_ENOMEM;
+	if (!lkl_is_running()) {
+		avail = sizeof(lkl_virtio_devs) - (devs - lkl_virtio_devs);
+		num_bytes = snprintf(devs, avail, " virtio_mmio.device=%d@0x%lx:%d",
+				     mmio_size, (uintptr_t) dev->base, dev->irq);
+		if (num_bytes < 0 || num_bytes >= avail) {
+			lkl_put_irq(dev->irq, "virtio");
+			unregister_iomem(dev->base);
+			lkl_host_ops.mem_free(dev->queue);
+			return -LKL_ENOMEM;
+		}
+		devs += num_bytes;
+		dev->virtio_mmio_id = lkl_num_virtio_boot_devs++;
+	} else {
+		ret =
```

Comment:
There are several "Must be called before calling lkl_start_kernel" in lkl.h. Can you correct them?

Can I add virtio devs after lkl is running concurrently in multiple threads? At least seems register_iomem() is not thread safe. If not, can you make a comment somewhere to make it clear?